### ISSUE TITLE
Backend v2 - Refactor & Testing & Error Handling

### ISSUE THREAD

**Issue by andrelandgraf** [andrelandgraf/high5-learning-backend#36]
As you can see, we only have 3 issues left in the backend. Yay! The backend now runs in most of the cases but it is not yet very stable. 

1. we received feedback for our first prototype and one big topic was nested promises in the backend. We have to rework nested mongoose calls to promise chaining. 

2. connected with the first bullet point: error handling. Promise chaining allows to throw Errors in every mongoose call and catch them later on. This allows nice error processing. We should throw precise errors and handle every possible failure. If we do this the backend stops to break occasionally.

3. Testing: we have not yet included any tests. I do not understand why the SEBA Team has not told us to do test driven design at least for the backend.  This would mean a lot of work for us now but we should at least talk about it: using a unit test framework for nodejs / express and implementing at least a few crucial tests.

**Comment by andrelandgraf** [comment]
3. could also help us that we do not break the backend while refactoring the mongoose calls to promise chains. 

Best example: we integrated school name selection in the registration process. @tum-aweink wanted the school to be part of the token and I thought I implemented is. The reason this confused me later on was because I have not tested the token in the first place and then it took @tum-aweink some time to find the error for me.
